### ISSUE TITLE
Cluster may become unavailable if a node goes down during bootstrap of another node

### ISSUE THREAD

**Issue by glommer** [scylladb/scylla#4488]
Problem reproduces 100% of the time.

If a node needed to stream ranges is down, bootstrap will not start, and it will print a message like this:

```
May 08 19:55:57 ip-10-2-151-198.ec2.internal scylla[7348]:  [shard 0] seastar - Exiting on unhandled exception: std::runtime_error (A node required to move the data consistently is down (10.2.161.183).  If you wish to move the data from a potentially inconsistent replica, restart the node with consistent_rangemovement=false)
```

However, if we start the bootstrap, and *then* in the middle of it a node fails, the cluster will become unavailable.

To understand why, it is important to realize that during bootstrap of a new node, the consistency level requirement of QUORUM is increased. Now we need the bootstrapping node to acknowledge the request as well.

If a node dies, and stay dead for long enough, bootstrapping node will fail its streams, and retry with increasing timeouts:

```
May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] stream_session - [Stream #b578fbb0-71cd-11e9-802c-000000000003] Fail to send PREPARE_MESSAGE to 10.2.161.183:0, seastar::rpc::closed_error (connection is closed)
May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] stream_session - [Stream #b578fbb0-71cd-11e9-802c-000000000003] Streaming error occurred, peer=10.2.161.183
May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] rpc - client 10.2.161.183: fail to connect: Connection refused
May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] stream_session - [Stream #b578fbb0-71cd-11e9-802c-000000000003] Streaming plan for Bootstrap-keyspace1-index-0 failed, peers={10.2.161.183}, tx=0 KiB, 0.00 KiB/s, rx=0 KiB, 0.00 KiB/s
May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] range_streamer - Bootstrap with 10.2.161.183 for keyspace=keyspace1 failed, took 0.011 seconds: streaming::stream_exception (Stream failed)
May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] range_streamer - Bootstrap failed, took 0 seconds, nr_ranges_remaining=257
May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] range_streamer - Bootstrap failed to stream. Will retry in 303 seconds ...
```

But if the node really doesn't come back, then bootstrap will fail :

```
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] boot_strapper - Error during bootstrap: streaming::stream_exception (Stream failed)
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] seastar - Exiting on unhandled exception: streaming::stream_exception (Stream failed)
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] compaction_manager - Asked to stop
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] compaction_manager - Stopped
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 3] compaction_manager - Asked to stop
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 1] compaction_manager - Asked to stop
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 1] compaction_manager - Stopped
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 3] compaction_manager - Stopped
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 2] compaction_manager - Asked to stop
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 2] compaction_manager - Stopped
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 1] storage_proxy - Failed to apply mutation from 10.2.13.122#1: seastar::gate_closed_exception (gate closed)
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 3] storage_proxy - Failed to apply mutation from 10.2.13.122#2: seastar::gate_closed_exception (gate closed)
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 3] storage_proxy - Failed to apply mutation from 10.2.13.122#2: seastar::gate_closed_exception (gate closed)
May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 1] storage_proxy - Failed to apply mutation from 10.2.13.122#1: seastar::gate_closed_exception (gate closed)
May 08 20:18:17 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] rpc - client 10.2.161.183: fail to connect: Connection refused
May 08 20:18:18 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 2] seastar - Exceptional future ignored: seastar::timed_out_error (timedout), backtrace:   0x7efeb8
```

However, note that it fails with an exception and exits right away. It doesn't seem to inform the remaining nodes that it exited, since they still think it is UJ:

```
Datacenter: us-east
===================
Status=Up/Down
|/ State=Normal/Leaving/Joining/Moving
--  Address	  Load       Tokens	  Owns    Host ID                               Rack
UN  10.2.55.82    10.79 GB   256          ?	  7cb902dd-42df-4078-89bb-6c334e666095  1f
UJ  10.2.151.198  ?          256          ?	  9450a188-e81c-4471-a6bd-cb7228458700  1f
DN  10.2.161.183  3.84 GB    256          ?	  401bd79d-d227-46db-ab0a-b2a85ab0ac43  1f
UN  10.2.13.122   10.53 GB   256          ?	  04df6314-00e6-4663-834e-de7f5b0b252c  1f
```

At this point it is no longer able to ack requests, and no writes can achieve the expanded consistency level.

**Comment by asias** [comment]
The other existing nodes will notice the bootstrap node is gone in ring_delay time and remove it from gossip.

**Comment by glommer** [comment]
What is the ring delay?
This is still unacceptable, because between the failure and ring_delay the cluster will be unavailable. We have to announce we're giving up before shutting down

**Comment by asias** [comment]
ring_delay_ms is default to 30s

**Comment by glommer** [comment]
this definitely failed then. I could still see the node as UJ hours later

**Comment by glommer** [comment]
@asias it is likely based on the above that ring_delay-based detection only works for UN nodes. Could that be the case ?

**Comment by vladzcloudius** [comment]
@asias And even if it takes 30s (or even a second) it's too much - during that time the cluster is going to be unavailable and requests are going to fail with no good reason.

The sequence should be something like the following (nodeA is a bootstrapping node):
1) NodeA fails the bootstrapping.
2) NodeA announces that it's going to be DOWN similar to how we announce a SHUTDOWN today.
3) Only after that is stops handling new writes and shuts itself down.

**Comment by avikivity** [comment]
As far as I can tell there are two choices: fail the write, or fail the bootstrap. Failing the write acknowleges that we had two failures with RF=4, and failing the bootstrap restores us to one failure with RF=3. Resumable bootstrap can help, but over-eager failing bootstrap is dangerous too.

**Comment by tgrabiec** [comment]
I think it's wrong for our algorithms to consider bootstrapping to be a
case of (RF + 1) for achieving CL. What we want to achieve is to make sure
that CL will be satisfied, as in there will be CL replicas having the
write, for rings both before and after the bootstrap. After the new node
takes ownership of some token range, one of the existing replicas will lose
ownership of that range (call it the yielding replica). The fact that the
yielding replica ACKed the write means nothing for achieving CL in the ring
after the bootstrap. That replica could be part of the range which was
already streamed to the bootstrapping node, so bootstrapping will not
replicate that write from the yielding replica to the bootstrapping node.
Once bootstrapping completes, the write on the yielding replica will be
gone. So I think what should happen is that we should consider two sets of
replicas, before the bootstrap, and after the bootstrap, both with the same
RF, and require to achieve CL in both sets.

For example, assume that we have RF=3 and the following replicas owning
some key: A, B and C. We are bootstrapping D, which will replace A once the
boostrapping completes. If both A and D are down, we can still achieve
CL=QUORUM in both {A, B, C} and {B, C, D}, so there's no reason to fail the
write. If D resumes and the bootstrapping finishes, CL will still be met.

Also, using RF+1 during bootstrap is wrong because it may result in CL not
being satisfied after bootstrap completes if RF is even. Consider RF=4 and
{A, B, C, D} + {E} (E=bootstrapping replacing A). If we treat this as RF=5,
then CL=QUORUM needs 3 nodes. So it could be satisfied from {A, B, C}. But
after bootstrap, the replica set is {B, C, D, E}. In that set, only 2
replicas have the write, and CL is not met.

**Comment by glommer** [comment]
I don't think you fully understood the issue Avi.

The problem happens when the bootstrap already failed. While the bootstrap
is still going on there are no failures.

**Comment by glommer** [comment]
I think Tomasz' suggestion is very good as far as how to handle the
increased RF.

(But I still think that announcing the shutdown properly would be helpful)

**Comment by asias** [comment]
> @asias And even if it takes 30s (or even a second) it's too much - during that time the cluster is going to be unavailable and requests are going to fail with no good reason.
> 
> The sequence should be something like the following (nodeA is a bootstrapping node):
> 
>     1. NodeA fails the bootstrapping.
> 
>     2. NodeA announces that it's going to be DOWN similar to how we announce a SHUTDOWN today.
> 
>     3. Only after that is stops handling new writes and shuts itself down.

What if the bootstrap node is a sudden shutdown (kill -9 the node), there is no chance the bootstrapping node could announce anything.  We need to handle this case as well.

**Comment by tgrabiec** [comment]
Having other nodes notice that bootstrapping is aborted sooner would help with availability in cases there is no network partition. 

If there is a network partition, we may not be able to achieve CL in the after-bootstrap replica set and cannot decide to abort the bootstrap unilaterally. There has to be consensus on this. Currently, only the bootstrapping node can decide this.

Another thing is that if streaming used QUORUM replicas as it should, then it wouldn't fail when one of the replicas is down. It would be even enough to fall back to another replica when one fails. Right now it is pretty much guaranteed to fail with one replica down.

**Comment by tgrabiec** [comment]
That sounds wrong, can you point to the code which implements that?

**Comment by asias** [comment]
> That sounds wrong, can you point to the code which implements that?

Can you explain why it is wrong?

Bootstrap node is not part of the ring. If the bootstrap node is not updating gossip status, the other node assumes it is gone.

The code is in: gossiper::do_status_check().

**Comment by tgrabiec** [comment]
On Thu, May 9, 2019 at 11:27 AM Asias He <notifications@github.com> wrote:

> That sounds wrong, can you point to the code which implements that?
>
> Can you explain why it is wrong?
>
> Bootstrap node is not part of the ring. If the bootstrap node is not
> updating gossip status, the other node assumes it is gone.
>
It's wrong for it to assume that it's gone. It could be just a network
partition and the bootstrapping node is not really gone and will become a
member of the ring when the partition heals.

**Comment by vladzcloudius** [comment]
> > @asias And even if it takes 30s (or even a second) it's too much - during that time the cluster is going to be unavailable and requests are going to fail with no good reason.
> > The sequence should be something like the following (nodeA is a bootstrapping node):
> > ```
> > 1. NodeA fails the bootstrapping.
> > 
> > 2. NodeA announces that it's going to be DOWN similar to how we announce a SHUTDOWN today.
> > 
> > 3. Only after that is stops handling new writes and shuts itself down.
> > ```
> 
> What if the bootstrap node is a sudden shutdown (kill -9 the node), there is no chance the bootstrapping node could announce anything. We need to handle this case as well.

@asias True. But in that case we'd have at least some excuse. When the bootstrapping node gracefully fails after 5 retries (@avikivity ) I don't think there is any excuse for us to fail any WRITE with CL=QUORUM.

**Comment by vladzcloudius** [comment]
> Having other nodes notice that bootstrapping is aborted sooner would help with availability in cases there is no network partition.
> 
> If there is a network partition, we may not be able to achieve CL in the after-bootstrap replica set and cannot decide to abort the bootstrap unilaterally. There has to be consensus on this. Currently, only the bootstrapping node can decide this.
> 
> Another thing is that if streaming used QUORUM replicas as it should, then it wouldn't fail when one of the replicas is down. It would be even enough to fall back to another replica when one fails. Right now it is pretty much guaranteed to fail with one replica down.

@tgrabiec If there is network partitioning and QUORUM (in a regular sense) may not be achieved then WRITE may fail - this is expected.

However I like the idea of streaming from the QUORUM of replicas. I even think this has been discussed at some point in the past.

**Comment by glommer** [comment]
@tgrabiec as Vlad correctly pointed out to me, there is a problem in your idea of requiring two disjoint sets

```
For example, assume that we have RF=3 and the following replicas owning
some key: A, B and C. We are bootstrapping D, which will replace A once the
boostrapping completes. If both A and D are down, we can still achieve
CL=QUORUM in both {A, B, C} and {B, C, D}, so there's no reason to fail the
write. If D resumes and the bootstrapping finishes, CL will still be met.
```

This actually does not solve the problem we saw here (not by itself anyway)

We saw a problem in which B died, and then D, as a result of the bootstrap failing, died as well.
In that case we will not be able to achieve a QUORUM in the second set, because B and D will not respond.

The bootstrap will have failed by then, that's true, but as we saw here, that information is not communicated immediately and until it is we will be unavailable. 

In that sense, @slivne idea seems slighly superior: if we attribute 0.5 to each A and D and 1 to B and C, we still can always achieve quorum with B and C.

**Comment by glommer** [comment]
@slivne idea won't solve this case either: if B and D fail (bootstrap fails), A and C will at most sum up to 1.5.

There is no solution to conserve availability that I can see that doesn't involve D announcing that it failed and being removed from the set ASAP before it can stop acking writes.

If D fails abruptly, I don't see how we can preserve availability in this case. But we may not need to: we are not safe against two hardware failures for QUORUM, and that is what would've happened in this case.

**Comment by tgrabiec** [comment]
On Thu, May 9, 2019 at 4:07 PM Glauber Costa <notifications@github.com>
wrote:

> @tgrabiec <https://github.com/tgrabiec> as Vlad correctly pointed out to
> me, there is a problem in your idea of requiring two disjoint sets
>
> For example, assume that we have RF=3 and the following replicas owning
> some key: A, B and C. We are bootstrapping D, which will replace A once the
> boostrapping completes. If both A and D are down, we can still achieve
> CL=QUORUM in both {A, B, C} and {B, C, D}, so there's no reason to fail the
> write. If D resumes and the bootstrapping finishes, CL will still be met.
>
> This actually does not solve the problem we saw here (not by itself anyway)
>
I didn't claim that my idea solves this problem, just that our current
algorithm is in some cases wrong, and in some cases too restrictive.

> We saw a problem in which B died, and then D, as a result of the bootstrap
> failing, died as well.
> In that case we will not be able to achieve a QUORUM in the second set,
> because B and D will not respond.
>
> The bootstrap will have failed by then, that's true, but as we saw here,
> that information is not communicated immediately and until it is we will be
> unavailable.
>
> In that sense, @slivne <https://github.com/slivne> idea seems slighly
> superior: if we attribute 0.5 to each A and D and 1 to B and C, we still
> can always achieve quorum with B and C.
>

**Comment by kostja** [comment]
* Tomasz Grabiec <notifications@github.com> [19/05/10 07:41]:

> I think it's wrong for our algorithms to consider bootstrapping to be a
> case of (RF + 1) for achieving CL. What we want to achieve is to make sure
> that CL will be satisfied, as in there will be CL replicas having the
> write, for rings both before and after the bootstrap. After the new node
> takes ownership of some token range, one of the existing replicas will lose
> ownership of that range (call it the yielding replica). The fact that the
> yielding replica ACKed the write means nothing for achieving CL in the ring
> after the bootstrap. That replica could be part of the range which was
> already streamed to the bootstrapping node, so bootstrapping will not
> replicate that write from the yielding replica to the bootstrapping node.
> Once bootstrapping completes, the write on the yielding replica will be
> gone. So I think what should happen is that we should consider two sets of
> replicas, before the bootstrap, and after the bootstrap, both with the same
> RF, and require to achieve CL in both sets.

Coincidentally, this is how Raft group membership change is
performed.  So +1.


-- 
Konstantin Osipov, Moscow, Russia, +7 903 626 22 32

**Comment by avikivity** [comment]
> On Thu, May 9, 2019 at 11:27 AM Asias He ***@***.***> wrote: That sounds wrong, can you point to the code which implements that? Can you explain why it is wrong? Bootstrap node is not part of the ring. If the bootstrap node is not updating gossip status, the other node assumes it is gone.
> It's wrong for it to assume that it's gone. It could be just a network partition and the bootstrapping node is not really gone and will become a member of the ring when the partition heals.


I can see two sides to this. If a bootstrapping node isn't making progress, it's preventing other nodes from bootstrapping (since we only allow one at a time).

On the other hand, since this is an administrative operation, we could say that the admin should eject that node, and if it reappears we could resume bootstrap (using repair-based bootstrap).

**Comment by glommer** [comment]
@avikivity About the general statement that once we can't get enough acks from both sets, we have to sacrifice either availability or the bootstrap, I've been thinking and this is likely a choice that we should push to the user, no ? (with a default to availability)

Some users will prefer the bootstrap to fail so they can try later (with resumable bootstraps this shouldn't even be a pain), and some users will prefer the bootstrap to continue and have decreased consistency.

**Comment by dorlaor** [comment]
On Fri, May 10, 2019 at 3:02 AM Glauber Costa <notifications@github.com>
wrote:

> @avikivity <https://github.com/avikivity> About the general statement
> that once we can't get enough acks from both sets, we have to sacrifice
> either availability or the bootstrap, I've been thinking and this is likely
> a choice that we should push to the user, no ? (with a default to
> availability)
>
> Some users will prefer the bootstrap to fail so they can try later (with
> resumable bootstraps this shouldn't even be a pain), and some users will
> prefer the bootstrap to continue and have decreased consistency.
>

Agree, we can even have a silly heuristic that compares the time
bootstraping and make re-try time equal to it until it gives up. When we
give up,
make sure we log it nicely.

**Comment by asias** [comment]
> Problem reproduces 100% of the time.
> 
> If a node needed to stream ranges is down, bootstrap will not start, and it will print a message like this:
> 
> ```
> May 08 19:55:57 ip-10-2-151-198.ec2.internal scylla[7348]:  [shard 0] seastar - Exiting on unhandled exception: std::runtime_error (A node required to move the data consistently is down (10.2.161.183).  If you wish to move the data from a potentially inconsistent replica, restart the node with consistent_rangemovement=false)
> ```
> 
> However, if we start the bootstrap, and _then_ in the middle of it a node fails, the cluster will become unavailable.
> 
> To understand why, it is important to realize that during bootstrap of a new node, the consistency level requirement of QUORUM is increased. Now we need the bootstrapping node to acknowledge the request as well.
> 
> If a node dies, and stay dead for long enough, bootstrapping node will fail its streams, and retry with increasing timeouts:
> 
> ```
> May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] stream_session - [Stream #b578fbb0-71cd-11e9-802c-000000000003] Fail to send PREPARE_MESSAGE to 10.2.161.183:0, seastar::rpc::closed_error (connection is closed)
> May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] stream_session - [Stream #b578fbb0-71cd-11e9-802c-000000000003] Streaming error occurred, peer=10.2.161.183
> May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] rpc - client 10.2.161.183: fail to connect: Connection refused
> May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] stream_session - [Stream #b578fbb0-71cd-11e9-802c-000000000003] Streaming plan for Bootstrap-keyspace1-index-0 failed, peers={10.2.161.183}, tx=0 KiB, 0.00 KiB/s, rx=0 KiB, 0.00 KiB/s
> May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] range_streamer - Bootstrap with 10.2.161.183 for keyspace=keyspace1 failed, took 0.011 seconds: streaming::stream_exception (Stream failed)
> May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] range_streamer - Bootstrap failed, took 0 seconds, nr_ranges_remaining=257
> May 08 20:13:13 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] range_streamer - Bootstrap failed to stream. Will retry in 303 seconds ...
> ```
> 
> But if the node really doesn't come back, then bootstrap will fail :
> 
> ```
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] boot_strapper - Error during bootstrap: streaming::stream_exception (Stream failed)
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] seastar - Exiting on unhandled exception: streaming::stream_exception (Stream failed)
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] compaction_manager - Asked to stop
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] compaction_manager - Stopped
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 3] compaction_manager - Asked to stop
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 1] compaction_manager - Asked to stop
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 1] compaction_manager - Stopped
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 3] compaction_manager - Stopped
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 2] compaction_manager - Asked to stop
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 2] compaction_manager - Stopped
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 1] storage_proxy - Failed to apply mutation from 10.2.13.122#1: seastar::gate_closed_exception (gate closed)
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 3] storage_proxy - Failed to apply mutation from 10.2.13.122#2: seastar::gate_closed_exception (gate closed)
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 3] storage_proxy - Failed to apply mutation from 10.2.13.122#2: seastar::gate_closed_exception (gate closed)
> May 08 20:18:16 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 1] storage_proxy - Failed to apply mutation from 10.2.13.122#1: seastar::gate_closed_exception (gate closed)
> May 08 20:18:17 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 0] rpc - client 10.2.161.183: fail to connect: Connection refused
> May 08 20:18:18 ip-10-2-151-198.ec2.internal scylla[7445]:  [shard 2] seastar - Exceptional future ignored: seastar::timed_out_error (timedout), backtrace:   0x7efeb8
> ```
> 
> However, note that it fails with an exception and exits right away. It doesn't seem to inform the remaining nodes that it exited, since they still think it is UJ:
> 
> ```
> Datacenter: us-east
> ===================
> Status=Up/Down
> |/ State=Normal/Leaving/Joining/Moving
> --  Address	  Load       Tokens	  Owns    Host ID                               Rack
> UN  10.2.55.82    10.79 GB   256          ?	  7cb902dd-42df-4078-89bb-6c334e666095  1f
> UJ  10.2.151.198  ?          256          ?	  9450a188-e81c-4471-a6bd-cb7228458700  1f
> DN  10.2.161.183  3.84 GB    256          ?	  401bd79d-d227-46db-ab0a-b2a85ab0ac43  1f
> UN  10.2.13.122   10.53 GB   256          ?	  04df6314-00e6-4663-834e-de7f5b0b252c  1f
> ```
> 
> At this point it is no longer able to ack requests, and no writes can achieve the expanded consistency level.

Let's get some basics cleared. 

@slivne  This is what I tried. So, after node4 gives up bootstrapping and exits immediately, other nodes will notice node4 is gone. 

node1, node2, node3 in the cluster
bootstrap node4
when node1,2,3 starts to streaming to node4, kill node2
node4 will continue to stream from node1,3 and fail to stream from node2 and retry and eventually give up the retry and exit
node1 and node3 will notice node4 is gone and remove node4 in ring delay time (30s) and clear node4's token from the pending token range.

@glommer Please confirm or deny that the write is recovered after existing nodes remove bootstrapping node.

**Comment by glommer** [comment]
Please note that Vlad sent a patch already for graceful exit. (it is in the review queue)
It would be nice to fix the problem of the node never being kicked out if it doesn't fail gracefully

What I want to achieve next with this is to give the user the choice of whether or not fail the bootstrap or fail the write in case of missing ack.

**Comment by slivne** [comment]
Glauber, yes we are trying to fix the original scenario in which the
bootrstapping node is not announcing shutdown and fails in the boostrap.

It seems Asias is not able to reproduce your issue

He is getting "node1 and node3 will notice node4 is gone and remove node4
in ring delay time (30s) and clear node4's token from the pending token
range."

While in your case (if I understood correctly) node1 and node3 do not
notice node4 went down - is that correct ?

Can you please comment on this

**Comment by glommer** [comment]
That is correct, in my test they simply never noticed node4 went down.
Is Asias trying this on 3.0 / 2019 ? If not, let's please do that. Maybe we fixed this already, and all we need is a backport.

**Comment by asias** [comment]
> That is correct, in my test they simply never noticed node4 went down.

How long did you wait to conclude this? Other nodes will not see node4 becomes DJ, instead you will see something like:
```
INFO  2019-05-14 17:56:17,924 [shard 0] gossip - FatClient 127.0.0.4 has been silent for 5000ms, removing from gossip  

```
After this, node4 is gone from the `nodetool status` output.

> Is Asias trying this on 3.0 / 2019 ? If not, let's please do that. Maybe we fixed this already, and all we need is a backport.

I tested on master. I did not see any patch that might change the behaviour from 3.0 to maser. The code is not changing in this area for a while.

**Comment by glommer** [comment]
I waited for many hours in my test setup, and the user also saw an 1h+ outage.

**Comment by asias** [comment]
> I waited for many hours in my test setup, and the user also saw an 1h+ outage.

Did you wait for "many hours" AFTER you see the bootstrapping exit? Note, the bootstrapping node will sleep and retry (5 times), sleep 60, 60*1.5, 60*1.5*1.5, ...

**Comment by glommer** [comment]
Before the bootstrapped node has exited, we don't become unavailable.

The cluster was unavailable for the user for 1h+.
And in my test yes, I controlled for that explicitly.

**Comment by vladzcloudius** [comment]
@asias @slivne AFAIU even 30s outage is unacceptable. I think it's even more than that - any write failure in this scenario is unacceptable because we are loosing only a single node from the cluster and CL=QUORUM should be satisfied.

So, under this definition using master I'm very easily able to reproduce the issue by doing exactly what Asias did but plus running a c-s write workload with cl=QUORUM (or LOCAL_QUORUM) at the same time.

My patch fixes that.

**Comment by glommer** [comment]
@slivne This is getting confusing, and both myself and Vlad got confused by it:
- losing writes is unacceptable (because the cluster is unavailable), and if we wait for the node to go away by its own we will have lost availability for 30s. Vlad sent a patch for that, it is still pending reviews.

 - You guys are right that in the case in which the node has a hard failure and does not have time to announce itself then the cluster has no choice but to become unavailable and gossip should boot it. If Asias is talking specifically about this, I suggest we open a new issue about this.

this issue would be solved by Vlad's patch because bootstrap failures, unlike hardware failures, give us the opportunity for graceful shutdown.

**Comment by slivne** [comment]
Yes asias and me are looking to solve the second case (the bootstraping
node fails and will not be able to announce)

Part of the problem is that asias was not able to reproduce

**Comment by glommer** [comment]
@slivne in this case maybe let's please open another issue to avoid confusion. This one will be auto-closed when Vlad's patches are in

**Comment by asias** [comment]
> Before the bootstrapped node has exited, we don't become unavailable.
> 
> The cluster was unavailable for the user for 1h+.
> And in my test yes, I controlled for that explicitly.

Glauber, you said "the cluster was unavailable for the user for 1h has the bootstrap node exited". But you did not mention if the other nodes saw the bootstrap was gone and removed from gossip. The unavailable might be something else other than the bootstrap node was not removed. 

Below is what I tried and I did not reproduce the issue bootstrap node was not removed by existing nodes.  

```
node1, node2, node3 in the cluster
bootstrap node4
when node1,2,3 starts to streaming to node4, kill node2
node4 will continue to stream from node1,3 and fail to stream from node2 and retry and eventually give up the retry and exit
node1 and node3 will notice node4 is gone and remove node4 in ring delay time (30s) and clear node4's token from the pending token range.
```

**Comment by glommer** [comment]
@asias you are trying this with master, though, no?
I did it with 3.0 / 2019. Maybe there is a bug that we ended up fixing in master.

I think @vladzcloudius looked at it a bit more already

**Comment by asias** [comment]
> It's wrong for it to assume that it's gone. It could be just a network partition and the bootstrapping node is not really gone and will become a member of the ring when the partition heals.

If there was a network partition, the bootstrap will fail because it fails to stream from that node.

**Comment by asias** [comment]
> @asias you are trying this with master, though, no?
> I did it with 3.0 / 2019. Maybe there is a bug that we ended up fixing in master.
> 
> I think @vladzcloudius looked at it a bit more already

I can try 3.0, but I highly doubt it will reproduce on 3.0. 

If you have logs for the run you saw the issue, it will be very easy to see if bootstrapping node is removed or not.

**Comment by asias** [comment]
> > @asias you are trying this with master, though, no?
> > I did it with 3.0 / 2019. Maybe there is a bug that we ended up fixing in master.
> > I think @vladzcloudius looked at it a bit more already
> 
> I can try 3.0, but I highly doubt it will reproduce on 3.0.
> 
> If you have logs for the run you saw the issue, it will be very easy to see if bootstrapping node is removed or not.

I tried with 3.0 branch. It did not reproduce. The bootstrapping node is removed as expected.

**Comment by slivne** [comment]
@vladzcloudius help - can you provide a scenario for Asias with 2019.1.4 / 3.2 that reproduces this - if not we will close.

**Comment by vladzcloudius** [comment]
@slivne I don't have any other scenario to propose except for the one described in the opening message.

If it has been fixed - I see no reason not to close this issue.

**Comment by slivne** [comment]
@asias does repair based operations behave any differently - that is, in a cluster that has a bootstrapping node starting and has another node in the cluster failing - will the bootstrapping node fail itself in a correct manner, can you test this out ?

do you know if we have a dtest with this scenario - if not create an issue and we will have someone code this scenario.

**Comment by asias** [comment]
Repair based should behave the same in terms of the increased RF to the bootstrap node. 

We can add a test like: 

Add n1,n2
Create ks with RF =2
Insert data with CL = 2
Bootstrap n3
Kill n3  before n3  finishes bootstrap
Check n1 and n2 will notice n3 is gone
Check writes with CL = 2 will recover

Notes: 
Write enough data so that we have enough time to kill n3 before it finishes
Some of the writes after n3 is killed will fail because it can not achieve the boosted consistency level during bootstrap.
The writes should recover once n1 and n2 think n3 is not bootstrapping.

**Comment by vladzcloudius** [comment]
> Repair based should behave the same in terms of the increased RF to the bootstrap node.
> 
> We can add a test like:
> 
> Add n1,n2
> Create ks with RF =2
> Insert data with CL = 2
> Bootstrap n3
> Kill n3 before n3 finishes bootstrap
> Check n1 and n2 will notice n3 is gone
> Check writes with CL = 2 will recover
> 
> Notes:
> Write enough data so that we have enough time to kill n3 before it finishes
> Some of the writes after n3 is killed will fail because it can not achieve the boosted consistency level during bootstrap.
> The writes should recover once n1 and n2 think n3 is not bootstrapping.

@slivne @roydahan Keep in mind that AFAIU by "kill" @asias means "kill -9" kill.
The idea is not to let `n3` inform other nodes that it went down.

If we want to have a test that checks the graceful `n3` shutdown we may want to add some debug scylla parameter that would shorten `bootstrapping` retries or (maybe even better) disable them at all.

After that we may test the case as originally described in the opening message by glauber.

**Comment by fruch** [comment]
we are working on a test-case to cover two case, killing gracefully and killing ungracefully, while bootstraping
 
our observation so far:
* **gracefully** - 
  **master** - the cluster seemed available from c-s POV for the whole time, no delays or failures (like observed in ungraceful test)
  **2019.1.9** - availability was lost, and c-s failed completely 

* **ungracefully** - both on 2019.1.9 and master, the cluster is unavailable only for the duration of `ring_delay_ms` and then c-s can continue.

**Comment by bhalevy** [comment]
Need to retest this with repair-based node operations and raft consistent topology changes.

Cc @kostja

**Comment by bhalevy** [comment]
> Need to retest this with repair-based node operations and raft consistent topology changes.
> 
> We already have 2 dtests that depend on this issue: test_cluster_become_unavailable_when_force_kill_node_during_bootstrap and test_cluster_become_unavailable_when_gracefully_kill_node_during_bootstrap
> 
> Cc @kostja

The tests fail due to cassandra-stress failing with:
```
E           java.io.IOException: Operation x10 on key(s) [4f4c4b34343939323431]: Error executing: (UnavailableException): Not enough replicas available for query at consistency TWO (2 required but only 1 alive)
```

**Comment by bhalevy** [comment]
> > Need to retest this with repair-based node operations and raft consistent topology changes.
> > We already have 2 dtests that depend on this issue: test_cluster_become_unavailable_when_force_kill_node_during_bootstrap and test_cluster_become_unavailable_when_gracefully_kill_node_during_bootstrap
> > Cc @kostja
> 
> The tests fail due to cassandra-stress failing with:
> 
> ```
> E           java.io.IOException: Operation x10 on key(s) [4f4c4b34343939323431]: Error executing: (UnavailableException): Not enough replicas available for query at consistency TWO (2 required but only 1 alive)
> ```

with `cl=TWO` and `rf=2` queries are expected to fail when the bootstrapping node is killed mid-way since we require to satisfy cl both on the old replica set and on the new (pending) set.
I'll adjust the test accordingly.

This issue can be closed after node operations correctly abort the operation when the bootstrapping node fails mid-way.

**Comment by bhalevy** [comment]
I understand from @tgrabiec that raft topology changes don't handle killing a bootstrapping node as gracefully and manual recovery is requried.  But this should be tracked using a separate GH issue.

Cc @gleb-cloudius
